keps/sig-network/3685-endpointslice-reconciler-to-staging/README.md

```diff
@@ -0,0 +1,154 @@
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Move the EndpointSlice reconciler and it’s dependencies to staging so that the
+reconciler logic be reused by out-of-tree EndpointSlice controllers.
+
+Some changes are needed in `pkg/controller/endpointslice` to move the reconciler
+code to `staging/src/k8s.io/endpointslice` and then import it as Go module.
+These changes include making private methods public and updating the import
+paths.
+
+## Motivation
+
+Moving the EndpointSlice reconciler to staging has several benefits:
+
+1. The reconciler logic can be reused by custom EndpointSlice controllers. The
+   fanout behavior is difficult to implement correctly and the current
+   reconciler code has gone through several iterations to reach its currently
+   level of robustness. There are cases where reconciler code has either been
+   forked or rewritten, both of which are problematic from a supportability
+   perspective.
+2. It reduces the burden of fully migrating from Endpoints to EndpointSlices.
+   It’s relatively easy to write a custom Endpoints controller, but less so with
+   EndpointSlices. So, the migration path is unclear for pre-existing custom
+   Endpoints controllers. Having a library that handles the more difficult
+   aspects of managing EndpointSlices would mitigate some of the risk in
+   migrating these controllers.
+3. It would help to realize one of the design goals of EndpointSlices: that
+   other controllers would exist to manage EndpointSlices, which is why the
+   `endpointslice.kubernetes.io/managed-by` annotation was added.
+
+### Goals
+
+- Expose a simple, and highly maintained, API for EndpointSlice reconciling.
+- Allow for the EndpointSlice reconciling code to easily be imported into other
+  projects.
+
+### Non-Goals
+
+- A generic solution for reconciling fan-out style resources.
```

Comment:
I think a natural question when reviewing this would be if we're going to add some kind of config on Service to tell the in-tree EndpointSlice controller to ignore it. At this point the only thing we have is an empty selector which is not ideal for some cases. If we ever do that, it should be a separate KEP, I just think it would be worth noting that it's been considered and is out of scope for this KEP.

Reply:
Ack, added it as a non-goal. Probably shares similar motivations, but I agree it should be a separate KEP.